DEPR: Change boxplot return_type kwarg

TomAugspurger opened the pull request:
Part of https://github.com/pydata/pandas/issues/6581
Deprecation started in https://github.com/pydata/pandas/pull/7096

Changes the default value of `return_type` in DataFrame.boxplot
and DataFrame.plot.box from None to 'axes'.

TomAugspurger committed:
DEPR: Change boxplot return_type kwarg

Part of https://github.com/pydata/pandas/issues/6581
Deprecation started in https://github.com/pydata/pandas/pull/7096

Changes the default value of `return_type` in DataFrame.boxplot
and DataFrame.plot.box from None to 'axes'.

jorisvandenbossche commented:
Tests still need to be updated

jreback commented:
@TomAugspurger pls update

jreback commented:
@TomAugspurger can you update

TomAugspurger commented:
Still working on this. It looks I didn't properly deprecate the `df.boxplot(by=)` (or my docs are wrong). According to the [docs](http://pandas-docs.github.io/pandas-docs-travis/visualization.html) `by=` is supposed to return an ordered dict of whatever `return_type` is. But it's returning a numpy array instead. I'll work on this more tonight, sorry.

jreback commented:
@TomAugspurger if you can get to this would be great, otherwise push to 0.19 :<

TomAugspurger commented:
I’ll update tonight or tomorrow morning. Might want to push anyway since I’ve found a discrepancy between our documented behavior and actual behavior for `groupby.boxplot`


> On Feb 15, 2016, at 2:22 PM, Jeff Reback <notifications@github.com> wrote:
> 
> @TomAugspurger <https://github.com/TomAugspurger> if you can get to this would be great, otherwise push to 0.19 :<
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/pydata/pandas/pull/12216#issuecomment-184374754>.
>

jreback commented:
@TomAugspurger ok, we do have a little time, and this has been around for a while. I'll keep it here and move later (or you decide  its a real stumbling block)

TomAugspurger commented:
Well, here's a not so great thing. I think we'll have to keep the behavior of `return_type=None` not being the same as `return_type='axes'` when `by` is not None.

```python
# before
a = df.groupby('d').boxplot()  # OD[dict]  warned
b = df.boxplot()               # dict, warned
c = df.boxplot(by='d')         # [axes], didn't warn
c = df.boxplot(by='d', return_type='axes') # OD[axes]
```

and

```python
# after
a = df.groupby('d').boxplot()  # OD[axes]
b = df.boxplot()               # axes
c = df.boxplot(by='d')         # [axes]
c = df.boxplot(by='d', return_type='axes') # OD[axes]
```

I had hoped to just make the default 'axes', but since I din't warn in the 3rd example, I don't think we can change that. Are we OK with that? Here's the updated docs:


```rst
Plot functions return scalar or arrays of :class:`matplotlib Axes <matplotlib.axes.Axes>`.
In ``boxplot``, the return type can be controlled by the ``return_type``, keyword. The valid choices are ``{"axes", "dict", "both"}``.
``'axes'`` returns a single matplotlib axes.
``'dict'`` returns a dict of matplotlib artists, similar to the matplotlib boxplot function.
``'both'`` returns a named tuple of axes and dicts.
When ``by`` is not None, you get back an ``OrderedDict`` of whatever ``return_type`` is, unless ``return_type`` is just ``None``, in which case you get back an array of axes.
```

jorisvandenbossche commented on doc/source/visualization.rst:
```diff
@@ -456,25 +456,14 @@ columns:
-Finally, when calling boxplot on a :class:`Groupby` object, a dict of ``return_type``
+Plot functions return scalar or arrays of :class:`matplotlib Axes <matplotlib.axes.Axes>`.
+In ``boxplot``, the return type can be controlled by the ``return_type``, keyword. The valid choices are ``{"axes", "dict", "both"}``.
+``'axes'`` returns a single matplotlib axes.
```
can you keep those as a list?

jorisvandenbossche commented:
(maybe not fully on topic, but) Is there actually a reason we are using OrderDicts here ? For example why not a series? This would give both the key access of an OrderDict as the integer indexing of a an array

jorisvandenbossche commented:
On returning a series: that is also what `df.groupby('d').plot()` does, so it seems logical that `df.groupby('d').boxplot()` does the same?

jorisvandenbossche commented:
But more on topic: wouldn't it be possible to raise that warning now for the third case? (but keeps its behaviour as you proposed) So we can maybe in a future release align the behaviour fully.

jreback commented on doc/source/visualization.rst:
```diff
@@ -456,25 +456,17 @@ columns:
-Basically, plot functions return :class:`matplotlib Axes <matplotlib.axes.Axes>` as a return value.
-In ``boxplot``, the return type can be changed by argument ``return_type``, and whether the subplots is enabled (``subplots=True`` in ``plot`` or ``by`` is specified in ``boxplot``).
+Plot functions return scalar or arrays of :class:`matplotlib Axes <matplotlib.axes.Axes>`.
+In ``boxplot``, the return type can be controlled by the ``return_type``, keyword. The valid choices are ``{"axes", "dict", "both"}``. If the ``by`` argument is ``None``,
```
would add a warning box that this is changed in 0.18.0

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -991,6 +991,7 @@ Removal of prior version deprecations/changes
 - Removal of ``rolling_corr_pairwise`` in favor of ``.rolling().corr(pairwise=True)`` (:issue:`4950`)
 - Removal of ``expanding_corr_pairwise`` in favor of ``.expanding().corr(pairwise=True)`` (:issue:`4950`)
 - Removal of ``DataMatrix`` module. This was not imported into the pandas namespace in any event (:issue:`12111`)
+- Changed the default value for the ``return_type`` parameter for ``DataFrame.plot.box`` and ``DataFrame.boxplot`` from ``None`` to ``"axes"``. These methods will now return a matplotlib axes by default instead of a dictionary of artists (:issue:`6581`).
```
add a see here (to the visualization.rst) warning box.

jreback commented:
@TomAugspurger lgtm. just some small doc comments. ping when updated.

jreback commented:
lgtm. 

@jorisvandenbossche ?

TomAugspurger commented:
Forgot to send the post I had started last night,

I could deprecate the third case (`df.boxplot(by='d')` -> [axes]), but at this point I feel a bit defeated by our boxplot implementations, plural, since we have 3(!), and none is equivalent to any other. It'd be nice to have them all aligned but I kind of feel like I've done enough API breaking changes here, especially when people should just use seaborn.

If we aren't warning on that case, this is good to go.

jorisvandenbossche commented:
@TomAugspurger About my comment on using Series instead of OrderedDict, wouldn't that be a possible way to just change?

To be clear: the third case is `c = df.boxplot(by='d')`, it currently gives you an array of axes (if there are multiple columns) and ideally we would like that it returns a, OrderedDict of axes to be consistent with a default of 'axes'. 
Suppose we would return a Series of axes, wouldn't that cover most of the use cases for both cases? Or would a lot of people relied on the fact it was an actual numpy array it returned? I would think most use is accessing one of the axes with ``c[0]``, which would still work with a Series as return value.

TomAugspurger commented:
@jorisvandenbossche yes, that makes much more sense. It's not clear to me the best way to get to there though. Do we return a `Series` for `c = df.boxplot(by='d') ` now, or have another FutureWarning?

jorisvandenbossche commented:
To answer that question, we should have an idea to which extent it would break people's code. And that is of course difficult to guess, but I think in the case of `c = df.boxplot(by='d')` return an array of axes, if you do something with that return value, it will typically be accessing one of the axes to modify the subplot further with matplotlib. That case would keep working with returning a Series. 
Of course code like `isintance(c, np.ndarray)` would break, but would people do that?

The questions is also, do we return Series for the cases where now already OrderedDicts are returned? Again, the most logical usecase (accessing a value by its key) keeps working, but things might not.

TomAugspurger commented:
My thinking was similar, the most common operations upon receiving the array or dict of axes is to 1. do nothing, 2. select individual axes by position or key. Thankfully Series handles both of those.

I'll update tonight to return Series of `return_type` wherever we returned an OrderedDict / numpy array before. It'll be nice to have this in a somewhat consistent and stable state.

TomAugspurger committed:
API: Change faceted boxplot return_type

Aligns behavior of `Groupby.boxplot` and DataFrame.boxplot(by=.)
to return a Series.

sinhrks commented:
I think returning ``Series`` is OK if plot method returns ``Axes``. Needs to consider to change other plots than boxplot though.

But for subplots, we receive a dict (per Axes) of dict (per Artist kinds) which contains list (per Artists). Are we using ``Series`` to store this kind of non-structural data?

```
df = pd.DataFrame({'A': [1, 2] * 20,
                   'B': np.random.randn(40),
                   'C': np.random.randn(40)})
df.boxplot(by='A', return_type='dict')
# OrderedDict([('B',
#               {'boxes': [<matplotlib.lines.Line2D at 0x164cb748>,
#                 <matplotlib.lines.Line2D at 0x163c9e48>],
#                'caps': [<matplotlib.lines.Line2D at 0x164b75f8>,
#                 <matplotlib.lines.Line2D at 0x163ee898>,
#                 <matplotlib.lines.Line2D at 0x162d8d68>,
#                ...
```

TomAugspurger commented:
@sinhrks  Correct, in the last case we return a dict.

http://nbviewer.jupyter.org/gist/TomAugspurger/9b0fb2309c7ac17ecf15

Thoughts? I think Series are dict-like enough to cover the common use-cases, selecting individual items for modification, or iteration through and modifying each.

I'm going to read through your groupby plotting enhancements PR next, to see how this would interact with it. I hadn't thought about changing the return type for all groupby.plot methods.

jreback commented:
looks ok to me. @jorisvandenbossche ?

jreback commented:
@TomAugspurger @jorisvandenbossche status on this?

TomAugspurger commented:
I think the concern is about returning Series for `Groupby.boxplot` and but for any of the other Groubpy plotting methods.

jreback commented:
why? previously these returned an ``OrderedDict`` or ``numpy`` array right?

What is the objection?

jreback commented:
@TomAugspurger @jorisvandenbossche are we changing this? or defer to 0.19.0

TomAugspurger commented:
Sorry, I was AWOL all saturday and yesterday.

I don't feel great about changing the groupby.boxplot() to return a Series without changing `groupby.plot` as well. We originally put the boxplot deprecation in place to make in more consistent with the rest of the plotting stuff, doesn't sense to introduce more inconsistencies in the process :)

I really need to review @sinhrks' groupby plotting changes, but I won't have a chance for at least a week. I would suggest pushing till 0.19 and doing all the groupby plotting changes then.

jreback commented:
@TomAugspurger ok, moved to 0.19.0. Though if you do have something before next week that makes sense. ping.

jorisvandenbossche commented:
Master is now targeted for 0.19.0, so we can revisit this

jreback commented:
@TomAugspurger ?

TomAugspurger commented:
Just rebased to fix a merge conflict in the release notes, but the last travis run did pass.

Are people OK with changing the result of `Groupby.boxplot` and `DataFrame.plot.box(by=col)` to a Series (instead of a numpy array / OrderedDict)?

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -331,10 +331,10 @@ API changes
 - ``__setitem__`` will no longer apply a callable rhs as a function instead of storing it. Call ``where`` directly to get the previous behavior. (:issue:`13299`)
 - Passing ``Period`` with multiple frequencies to normal ``Index`` now returns ``Index`` with ``object`` dtype (:issue:`13664`)
 - ``PeriodIndex.fillna`` with ``Period`` has different freq now coerces to ``object`` dtype (:issue:`13664`)
+- Faceted boxplots (from ``DataFrame.groupby(col).boxplot`` or ``DataFrame.boxplot(by=col))`` now return a ``Series``. Previously the groupby version returned an ``OrderedDict``, while the by version returned a NumPy array (:issue:`12216`)
```
previously this returned a NumPy Array

jreback commented on pandas/tests/test_graphics.py:
```diff
@@ -420,22 +420,20 @@ def _check_box_return_type(self, returned, return_type, expected_keys=None,
         types = {'dict': dict, 'axes': Axes, 'both': tuple}
         if expected_keys is None:
-            # should be fixed when the returning default is changed
             if return_type is None:
```
this should be an error if ``return_type`` is not valid (e.g. ``axes``)? ``None`` is no longer acceptable.

jreback commented on pandas/tests/test_graphics.py:
```diff
@@ -420,22 +420,20 @@ def _check_box_return_type(self, returned, return_type, expected_keys=None,
                     self.assertIsInstance(r, Axes)
                 return
 
-            self.assertTrue(isinstance(returned, OrderedDict))
```
you can use ``assert_series_equal`` instead (it checks type)

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -331,10 +331,10 @@ API changes
 - Passing ``Period`` with multiple frequencies to normal ``Index`` now returns ``Index`` with ``object`` dtype (:issue:`13664`)
 - ``PeriodIndex.fillna`` with ``Period`` has different freq now coerces to ``object`` dtype (:issue:`13664`)
+- Faceted boxplots (from ``DataFrame.groupby(col).boxplot`` or ``DataFrame.boxplot(by=col))`` now return a ``Series``. Previously the groupby version returned an ``OrderedDict``, while the by version returned a NumPy array (:issue:`12216`)
 - More informative exceptions are passed through the csv parser. The exception type would now be the original exception type instead of ``CParserError``. (:issue:`13652`)
```
maybe reference the original issue as well (where it was deprecated)

jreback commented:
minor comments, lgtm.

sinhrks commented:
I'm still -1 to return ``Series`` because the original intention of #7096 is to make all plot functions to return Axes (or NumPy array of axes) consistently. Returning ``Series`` in doesn't meet the original requirement.

Though I'm not fully understand what is being returned actually... is below understanding correct ? May better to be in documented anywhere.

**normal ``.plot``**

|return_type|subplots=False|subplots=True|
|------------|--------------|---------------|
| not affected | axes | numpy array of axes |

**``.plot(box)`` and ``.boxplot`` (0.18.1)**

|return_type|subplots=False|subplots=True / by='...' |
|------------|--------------|---------------|
| None          | axes               | numpy array of axes |
| dict             | dict of artists | OrderedDict contains dict of artists |
| axes           | axes                | OrderedDict contains axes |
| both           | namedtuple   | OrderedDict contains namedtuple |

**``.plot(box)`` and ``.boxplot`` (this PR)**

|return_type|subplots=False|subplots=True / by='...' |
|------------|--------------|---------------|
| None          | axes               | Series contains axes |
| dict             | dict of artists | Series contains dict of artists |
| axes           | axes                | Series contains axes |
| both           | namedtuple   | Series contains namedtuple |

jorisvandenbossche commented:
@sinhrks But note that in your overview above, none of the 'axes' return values are changed to a series, only the ones that returned numpy arrays or OrderedDicts. So I don't think this does anything against the intention of always returning Axes?

TomAugspurger commented:
@sinhrks the default will still now an axes (enforcing the deprecation of return_type='dict' to return_type='axes'. The thing that changes is before:

- Groupby.boxplot: returns array of axes
- Series.boxplot: returned an OrderedDict of `return_type`

Now both of those return a Series of `return_type`. I did steal your table and add it to the latest commit though :)

jreback commented:
status of this?

jorisvandenbossche commented on pandas/tests/test_graphics_others.py:
```diff
@@ -310,8 +311,8 @@ def _check_ax_limits(col, ax):
         p = df.boxplot(['height', 'weight', 'age'], by='category')
-        height_ax, weight_ax, age_ax = p[0, 0], p[0, 1], p[1, 0]
-        dummy_ax = p[1, 1]
+        height_ax, weight_ax, age_ax = p[0], p[1], p[2]
```
This is a breaking change? So previously 2D-array, now 1D-Series?

TomAugspurger replied:
Apparently it is. I didn't quite realize that as I was getting the tests to pass :/

jorisvandenbossche commented:
So to summarize to my understanding. There are two separate changes in this PR:

- change of the default for `return_type` (from None to 'axes', as was previously warned for) -> this is OK, not controversial
- change grouped boxplots (by groupby.boxplot or df.boxplot(by=..) to return A series of axes (or of dicts, depending on `return_type`) instead of array or OrderedDict -> this is the more controversial change
  - `df.groupby().boxplot()` -> this returns currently an OderedDict, and changed in this PR to return a Series -> IMO not a problem (should not break code), and more consistent with `df.groupby().plot()` (this also returns a Series of ax objects)
  - `df.boxplot(by=...)` -> this returns currently an array of axes, and would be changed by this PR to return Series of axes. 
    - The issue I now see here, is that if you have more columns so that the axes do not fit on one row, a 2D array was returned, while the Series is of course 1D. This can break code ..

The table above of @sinhrks (https://github.com/pydata/pandas/pull/12216#issuecomment-234243296) is a correct understanding of the situation, but

> I'm still -1 to return Series because the original intention of #7096 is to make all plot functions to return Axes (or NumPy array of axes) consistently

note that all methods now *do* return axes by default, only in the grouped cases a Series of axes instead an array of axes. So the original intent of #7096 is certainly kept.

To conclude, I am +1 on merging this (although the specific case of 2D numpy array -> 2D Series is a bit annoying) 
@sinhrks your thoughts?

@TomAugspurger if we decide to merge, needs a rebase :-)

sinhrks commented:
I prefer the result below to handle supblot axes using its shape (plot layout).

|return_type|subplots=False|subplots=True / by='...' |
|------------|--------------|---------------|
| None          | axes               | numpy array of axes |
| dict             | dict of artists | Series contains dict of artists |
| axes           | axes                | numpy array of axes |
| both           | namedtuple   | Series contains namedtuple |

Though I don't against the merge, we can reconsider if any feedbacks:)

TomAugspurger commented:
I see the arguments for `by='column', return_type='axes'` returning a 2-D array of axes to match the layout of the subplots, rather than a Series of arrays of axes.

Are people OK with this?

```python
In [1]: df = sns.load_dataset('tips')  # has 3 numeric columns to plot

# default, return_type='axes', returns a 2x2 array of axes (1 blank)
In [3]: df.boxplot(by='day')  
Out[3]:
array([[<matplotlib.axes._subplots.AxesSubplot object at 0x10b7c57f0>,
        <matplotlib.axes._subplots.AxesSubplot object at 0x111169fd0>],
       [<matplotlib.axes._subplots.AxesSubplot object at 0x1111a3cc0>,
        <matplotlib.axes._subplots.AxesSubplot object at 0x1111d9588>]], dtype=object)

# return_type=dict -> Series with the groupby keys as the index
In [4]: df.boxplot(by='day', return_type='dict')
Out[4]:
size          {'boxes': [Line2D(_line0), Line2D(_line7), Lin...
tip           {'boxes': [Line2D(_line0), Line2D(_line7), Lin...
total_bill    {'boxes': [Line2D(_line0), Line2D(_line7), Lin...
dtype: object

# return_type='both' -> Series with groupby keys as the index
In [5]: df.boxplot(by='day', return_type='both')
Out[5]:
size          (Axes(0.1,0.559091;0.363636x0.340909), {'boxes...
tip           (Axes(0.536364,0.559091;0.363636x0.340909), {'...
total_bill    (Axes(0.1,0.15;0.363636x0.340909), {'boxes': [...
dtype: object
```

jorisvandenbossche commented:
The overview in the latest post is certainly an improvement and ok to merge for me!

Although I find the inconsistent return type (series vs array) depending on the value of `return_type` not really ideal ... (but changing from array to series is of course not ideal as well, given the breaking change from 2D array to 1D series in certain cases, so OK)

jorisvandenbossche commented on doc/source/visualization.rst:
```diff
@@ -456,28 +456,35 @@ columns:
-* if ``return_type`` is ``'both'`` a namedtuple containing the :class:`matplotlib Axes <matplotlib.axes.Axes>`
-   and :class:`matplotlib Lines <matplotlib.lines.Line2D>` is returned
+Faceted boxplots, created by ``Groupby.boxplot`` or ``DataFrame.boxplot`` with the ``by``
+keyword will return either a NumPy array of ``axes``, when ``return_type='axes'``,
```
This is only true in case of `DataFrame.boxplot`, not for `Groupby.boxplot` (unless you changed that)

jorisvandenbossche commented on doc/source/visualization.rst:
```diff
@@ -456,28 +456,35 @@ columns:
+Faceted boxplots, created by ``Groupby.boxplot`` or ``DataFrame.boxplot`` with the ``by``
+keyword will return either a NumPy array of ``axes``, when ``return_type='axes'``,
+or a ``Series`` of ``return_type``. Each column of the input ``DataFrame``
+will get a single facet.
```
This is also not true in case of `groupby.boxplot` (there the facets are the groups)

Given those difference, maybe only speak about `DataFrame.boxplot` here?

TomAugspurger replied:
Fixed. Made clear that the table only applies to `DataFrame.boxplot`, and noted that `Groupby.boxplot` always returns a Series of `return_type`.

jorisvandenbossche commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -437,6 +437,7 @@ API changes
 - ``__setitem__`` will no longer apply a callable rhs as a function instead of storing it. Call ``where`` directly to get the previous behavior. (:issue:`13299`)
 - Passing ``Period`` with multiple frequencies to normal ``Index`` now returns ``Index`` with ``object`` dtype (:issue:`13664`)
 - ``PeriodIndex.fillna`` with ``Period`` has different freq now coerces to ``object`` dtype (:issue:`13664`)
+- Faceted boxplots (from ``DataFrame.groupby(col).boxplot`` or ``DataFrame.boxplot(by=col))`` now return a ``Series``. Previously the groupby version returned an ``OrderedDict``, while the by version returned a NumPy Array (:issue:`12216`, :issue:`7096`)
```
... except for ``DataFrame.boxplot(by=col, return_type='axes')`` which still returns an array

jorisvandenbossche replied:
also "boxplot(by=col))" -> 1 paranthesis too much

TomAugspurger replied:
Fixed

> - Faceted boxplots from ``DataFrame.boxplot(by=col)`` now return a ``Series`` when ``return_type='dict'`` or ``return_type='both'``. Previously these returned an ``OrderedDict``. Note that when ``return_type='axes'``, the default, these still return a 2-D NumPy array. (:issue:`12216`, :issue:`7096`)

TomAugspurger commented:
Ok, this is probably more info than anyone cares about, but

The key one to note is the difference between `return_type=None` and `return_type='axes'` in `.boxplot(by='column'`). The first returns an array of axes, the second returns a Series / dict of axes.

method    | faceted  | keyword  | 0.18 output | 0.19 output  | notes     |
----------|----------|----------|-------------|--------------|-----------|
.plot.box | No       | None     | axes        | axes         |           |
.plot.box | No       | axes     | axes        | axes         |           |
.plot.box | No       | dict     | dict        | dict         |           |
.plot.box | No       | both     | both        | both         |           |
.plot.box | by       | None     | axes        | axes         | no facet  |
.plot.box | by       | axes     | axes        | axes         | no facet  |
.plot.box | by       | dict     | dict        | dict         | no facet  |
.plot.box | by       | both     | both        | both         | no facet  |
--------- | -------- | -------- | ----------- | ------------ | --------  |
.boxplot  | No       | None     | dict        | axes         | depr warn |
.boxplot  | No       | axes     | axes        | axes         |           |
.boxplot  | No       | dict     | dict        | dict         |           |
.boxplot  | No       | both     | both        | both         |           |
.boxplot  | by       | None     | axes        | axes         | no depr   |
.boxplot  | by       | axes     | ODict[axes] | Series[axes]        | !         |
.boxplot  | by       | dict     | ODict[dict] | Sereis[dict] |           |
.boxplot  | by       | both     | ODict[both] | Series[both] |           |
--------- | -------- | -------- | ----------- | ------------ | --------  |
.groupby  | yes      | None     | ODict[dict] | Series[axes] | depr warn |
.groupby  | yes      | axes     | ODict[axes] | Series[axes] |           |
.groupby  | yes      | dict     | ODict[dict] | Series[dict] |           |
.groupby  | yes      | both     | Odict[both] | Series[both] |           |
--------- | -------- | -------- | ----------- | ------------ | --------  |

Since we didn't deprecate the difference there, I'll leave it in place. Will get this all cleaned up tonight and submit for review. Sorry this has taken so long :/

jreback commented:
is there a reason we r not deprecating
.boxplot in favor of .plot.box

seems otherwise this is very confusing and not consistent

TomAugspurger commented:
Updated, Travis is green. To summarize the changes:

1. Enforce the deprecation on `.boxplot` and `.groupby.boxplot` return_type. Previously it was `'dict'`, now it's `'axes`'.
2. Return `Series` instead of `OrderedDict`s from faceted boxplots, which is more consistent with other plotting methods. The exception is `.boxplot(by='col')` with the default `return_type=None`, which continues to return a 2-D NumPy array

> seems otherwise this is very confusing and not consistent

Extremely 😞  I believe that after this deprecation we should have a clear path forward API-wise, unless I'm missing something.

jreback commented on doc/source/visualization.rst:
```diff
@@ -456,28 +456,39 @@ columns:
+
+Most plot functions return a scalar or an array of :class:`matplotlib Axes <matplotlib.axes.Axes>`.
+In ``boxplot``, the return type can be controlled by the ``return_type``, keyword. The valid choices are ``{"axes", "dict", "both"}``. If the ``by`` argument is ``None``,
+
```
I don't think you need this with the table below. IOW would rather see the table, then comment if needed on particular ones?

TomAugspurger commented:
FYI the only CI failure here was codecov.

jreback commented on pandas/tools/plotting.py:
```diff
@@ -2691,14 +2691,16 @@ def plot_series(data, kind='line', ax=None,                    # Series unique
-        When grouping with ``by``, a dict mapping columns to ``return_type``
-        is returned.
+        When grouping with ``by``, a Series mapping columns to ``return_type``
+        is returned, unless ``return_type`` is None, in which case a NumPy
```
maybe add a link to the docs here

jreback commented on pandas/tools/plotting.py:
```diff
@@ -3175,15 +3173,15 @@ def _grouped_plot_by_column(plotf, data, columns=None, by=None,
         ax.set_title(col)
         ax.set_xlabel(pprint_thing(by))
-        result[col] = re_plotf
+        result.loc[col] = re_plotf
```
this is completely non-performant (copies each time). should append to lists (data/index) and create at the end (or create the Series at the end from the OrderedDict might be better/easier)

jreback commented:
couple comments, but lgtm. merge on green. this has been outstanding for quite a while, let's just put it in.

jorisvandenbossche commented:
@TomAugspurger Thanks!
